x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/boxplot/InternalBoxplot.java

```diff
@@ -0,0 +1,200 @@
+            case max:
+                return getMax();
+            case q1:
+                return getQ1();
+            case q2:
+                return getQ2();
+            case q3:
+                return getQ3();
+            default:
+                throw new IllegalArgumentException("Unknown value [" + name + "] in common stats aggregation");
+        }
+    }
+
+    // for testing only
+    DocValueFormat format() {
+        return format;
+    }
+
+    // for testing only
+    TDigestState state() {
+        return state;
+    }
+
+    @Override
+    public InternalBoxplot reduce(List<InternalAggregation> aggregations, ReduceContext reduceContext) {
+        TDigestState merged = null;
+        for (InternalAggregation aggregation : aggregations) {
+            final InternalBoxplot percentiles = (InternalBoxplot) aggregation;
+            if (merged == null) {
+                merged = new TDigestState(percentiles.state.compression());
+            }
+            merged.add(percentiles.state);
+        }
+        return new InternalBoxplot(name, merged, format, pipelineAggregators(), metaData);
+    }
+
+    static class Fields {
+        public static final String MIN = "min";
+        public static final String MIN_AS_STRING = "min_as_string";
+        public static final String MAX = "max";
+        public static final String MAX_AS_STRING = "max_as_string";
+        public static final String Q1 = "q1";
+        public static final String Q1_AS_STRING = "q1_as_string";
+        public static final String Q2 = "q2";
+        public static final String Q2_AS_STRING = "q2_as_string";
+        public static final String Q3 = "q3";
+        public static final String Q3_AS_STRING = "q3_as_string";
+    }
+
+    @Override
+    public XContentBuilder doXContentBody(XContentBuilder builder, Params params) throws IOException {
+        builder.field(Fields.MIN, getMin());
+        builder.field(Fields.MAX, getMax());
+        builder.field(Fields.Q1, getQ1());
+        builder.field(Fields.Q2, getQ2());
+        builder.field(Fields.Q3, getQ3());
+        if (format != DocValueFormat.RAW) {
+            builder.field(Fields.MIN_AS_STRING, format.format(getMin()));
+            builder.field(Fields.MAX_AS_STRING, format.format(getMax()));
+            builder.field(Fields.Q1_AS_STRING, format.format(getQ1()));
+            builder.field(Fields.Q2_AS_STRING, format.format(getQ2()));
+            builder.field(Fields.Q3_AS_STRING, format.format(getQ3()));
+        }
+        return builder;
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(super.hashCode(), state);
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (this == obj) return true;
+        if (obj == null || getClass() != obj.getClass()) return false;
+        if (super.equals(obj) == false) return false;
+
+        InternalBoxplot that = (InternalBoxplot) obj;
+        return Objects.equals(state, that.state);
```

Comment:
Should we include `format` in this equality check?

Philosophically, I wonder if we should check equality of the boxplot components instead of the state? Like, are two boxplots identical if min/max/q1/q2/q3 are the same but the rest of the distribution is different?

 I really dunno heh :)

Reply:
It is already included in super.

Reply:
:man_facepalming: